src/ssl.c

```diff
@@ -258,6 +260,533 @@ const WOLF_EC_NIST_NAME kNistCurves[] = {
+            /* publicName */
+            XMEMCPY(workingConfig->publicName, echConfig, publicNameLen);
+            /* null terminated */
+            workingConfig->publicName[publicNameLen] = 0;
+
+            /* add length to go to next config, +4 for version and length */
+            i += length + 4;
+
+            /* check that we support this config */
+            for (j = 0; j < HPKE_SUPPORTED_KEM_LEN; j++) {
+                if (hpkeSupportedKem[j] == workingConfig->kemId)
+                    break;
+            }
+
+            /* if we don't support the kem or at least one cipher suite */
+            if (j >= HPKE_SUPPORTED_KEM_LEN ||
+              EchConfigGetSupportedCipherSuite(workingConfig) < 0) {
+                XFREE(workingConfig->cipherSuites, ssl->heap,
+                    DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(workingConfig->publicName, ssl->heap,
+                    DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(workingConfig->raw, ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
+                workingConfig = lastConfig;
+            }
+        } while ((word32)i < echConfigsLen);
+    }
+
+    /* if we found valid configs */
+    if ( configList != NULL ) {
+        ssl->options.useEch = 1;
+        ssl->echConfigs = configList;
+
+        return WOLFSSL_SUCCESS;
+    }
+
+    return WOLFSSL_FATAL_ERROR;
+}
+
+int wolfSSL_GetEchConfig(WOLFSSL_EchConfig* config, byte* output,
+    word32* outputLen)
+{
+    int i;
+    byte* outputStart = output;
+    word16 totalLen = 0;
+
+    if (config == NULL || (output == NULL && outputLen == NULL))
+        return BAD_FUNC_ARG;
+
+    /* 2 for version */
+    totalLen += 2;
+    /* 2 for length */
+    totalLen += 2;
+    /* 1 for configId */
+    totalLen += 1;
+    /* 2 for kemId */
+    totalLen += 2;
+    /* 2 for hpke_len */
+    totalLen += 2;
+
+    /* hpke_pub_key */
+    switch (config->kemId) {
+        case DHKEM_P256_HKDF_SHA256:
+            totalLen += DHKEM_P256_ENC_LEN;
+            break;
+        case DHKEM_P384_HKDF_SHA384:
+            totalLen += DHKEM_P384_ENC_LEN;
+            break;
+        case DHKEM_P521_HKDF_SHA512:
+            totalLen += DHKEM_P521_ENC_LEN;
+            break;
+        case DHKEM_X25519_HKDF_SHA256:
+            totalLen += DHKEM_X25519_ENC_LEN;
+            break;
+        case DHKEM_X448_HKDF_SHA512:
+            totalLen += DHKEM_X448_ENC_LEN;
+            break;
+    }
+
+    /* cipherSuitesLen */
+    totalLen += 2;
+    /* cipherSuites */
+    totalLen += config->numCipherSuites * 4;
+    /* public name len */
+    totalLen += 2;
+
+    /* public name */
+    totalLen += XSTRLEN(config->publicName);
+    /* trailing zeros */
+    totalLen += 2;
+
+    if (output == NULL) {
+        *outputLen = totalLen;
+        return LENGTH_ONLY_E;
+    }
+
+    if (totalLen > *outputLen) {
+        *outputLen = totalLen;
+        return INPUT_SIZE_E;
+    }
+
+    /* version */
+    c16toa(TLSX_ECH, output);
+    output += 4;
```

Comment:
Only put two bytes in. Add comment that 2 more bytes are added for the length that will be added at the end.
Is totalLen exactly the right amount? If so, could be use it to put the length in now?